pkg/adhocprofiles/adhocprofiles.go

```diff
@@ -0,0 +1,256 @@
+		UploadedAt:         adHocProfile.UploadedAt.UnixMilli(),
+		FlamebearerProfile: string(jsonProfile),
+		ProfileType:        profile.Metadata.Name,
+		ProfileTypes:       profileTypes,
+	}), nil
+}
+
+func (a *AdHocProfiles) Get(ctx context.Context, c *connect.Request[v1.AdHocProfilesGetRequest]) (*connect.Response[v1.AdHocProfilesGetResponse], error) {
+	tenantID, err := tenant.TenantID(ctx)
+	if err != nil {
+		return nil, connect.NewError(connect.CodeInvalidArgument, err)
+	}
+
+	bucket, err := a.getBucket(tenantID)
+	if err != nil {
+		return nil, err
+	}
+
+	reader, err := bucket.Get(ctx, c.Msg.GetId())
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get profile")
+	}
+
+	adHocProfileBytes, err := io.ReadAll(reader)
+	if err != nil {
+		return nil, err
+	}
+
+	var adHocProfile AdHocProfile
+	err = json.Unmarshal(adHocProfileBytes, &adHocProfile)
+	if err != nil {
+		return nil, err
+	}
+
+	maxNodes, err := validation.ValidateMaxNodes(a.limits, []string{tenantID}, c.Msg.GetMaxNodes())
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not determine max nodes")
+	}
+
+	profile, profileTypes, err := parse(&adHocProfile, c.Msg.ProfileType, maxNodes)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to parse profile")
+	}
+
+	jsonProfile, err := json.Marshal(profile)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to parse profile")
+	}
+
+	return connect.NewResponse(&v1.AdHocProfilesGetResponse{
+		Id:                 c.Msg.Id,
+		Name:               adHocProfile.Name,
+		UploadedAt:         adHocProfile.UploadedAt.UnixMilli(),
+		FlamebearerProfile: string(jsonProfile),
+		ProfileType:        profile.Metadata.Name,
+		ProfileTypes:       profileTypes,
+	}), nil
+}
+
+func (a *AdHocProfiles) List(ctx context.Context, c *connect.Request[v1.AdHocProfilesListRequest]) (*connect.Response[v1.AdHocProfilesListResponse], error) {
+	bucket, err := a.getBucketFromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	profiles := make([]*v1.AdHocProfilesProfileMetadata, 0)
+	err = bucket.Iter(ctx, "", func(s string) error {
+		separatorIndex := bytes.IndexByte([]byte(s), '-')
+		id, err := ulid.Parse(s[0:separatorIndex])
+		if err != nil {
+			level.Warn(a.logger).Log("msg", "cannot parse ad hoc profile", "key", s, "err", err)
+			return nil
+		}
+		name := s[separatorIndex+1:]
+		profiles = append(profiles, &v1.AdHocProfilesProfileMetadata{
+			Id:         s,
+			Name:       name,
+			UploadedAt: int64(id.Time()),
+		})
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+	return connect.NewResponse(&v1.AdHocProfilesListResponse{Profiles: profiles}), nil
+}
+
+func (a *AdHocProfiles) getBucketFromContext(ctx context.Context) (objstore.Bucket, error) {
+	tenantID, err := tenant.TenantID(ctx)
+	if err != nil {
+		return nil, connect.NewError(connect.CodeInvalidArgument, err)
+	}
+	return a.getBucket(tenantID)
+}
+
+func (a *AdHocProfiles) getBucket(tenantID string) (objstore.Bucket, error) {
```

Comment:
I don't think you need to cache buckets you could create one child per request and throw it away.

but may be there's a reason ?